flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala

```diff
@@ -103,27 +97,42 @@ class DataStreamJoin(
       tableEnv: StreamTableEnvironment,
       queryConfig: StreamQueryConfig): DataStream[CRow] = {
 
-    val config = tableEnv.getConfig
-    val returnType = schema.typeInfo
-    val keyPairs = joinInfo.pairs().toList
+    validateKeyTypes()
 
-    // get the equality keys
-    val leftKeys = ArrayBuffer.empty[Int]
-    val rightKeys = ArrayBuffer.empty[Int]
+    val leftDataStream =
```

Comment:
I would be against that. I'm depending on this change in this PR, so it defines strict order of those two commits. Either I wouldn't publish the actual versioned joins PR after this commit is merged, or the versioned joins PR would look exactly the same as it looks now. Splitting into two PRs adds a lot of overhead and room for a lot of problems with PRs going out of sync. During reviewing it's also annoying, because you see the same code twice and reviewers (same or different ones) are often (almost always?) making half of the comments in one PR and half in the other.

Besides, this is already in a separate commit so it can be reviewed separately. Maybe instead of reviewing PR all at once try reviewing it commit by commit?